text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+  // For a description of the Arrow IPC format see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+  bytes record = 4;
+
+  // [mandatory]
+  CompressionMethod compression = 5;
+}
+
+enum OtlpArrowPayloadType {
+  METRICS = 0;
+  LOGS = 1;
+  SPANS = 2;
+}
+
+message EncodedData {
+  bytes ipc_message = 1;
+  bytes arrow_data = 2;
+}
+
+enum CompressionMethod {
+  NO_COMPRESSION = 0;
+  ZSTD = 1;
+}
+```
+
+The `sub_stream_id` attribute is a unique identifier for a sub-stream of `BatchArrowRecords` sharing the same schema inside
+the
+scope of the current stream. This id will be used receiver side to keep track of the schema and dictionaries for a
+specific
+type of Arrow Records. See the [Substream Id generation](#substream-id-generation) section for more information on the
+implementation
+of this identifier.
+
+The `OtlpArrowPayloadType` enum specifies the `type` of the payload.
+
+The `record` attribute is a binary representation of the Arrow RecordBatch.
+
+The `compression` attribute is a mandatory attribute that is used to define the compression algorithms for the different
+bytes buffer.
+
+ZSTD offers a training mode, which can be used to tune the algorithm for a selected type of data. The result of this
+training is a dictionary that can be used to compress the data. Using this [dictionary](http://facebook.github.io/zstd/#small-data)
+can dramatically improve the compression rate for small batches. This future development will build on both the gRPC
+stream approach used in this proposal and the ability to send a ZSTD dictionary over the OTLP Arrow stateful protocol,
+allowing us to train the ZSTD algorithm on the first batches and then update the configuration of the ZSTD
+encoder/decoder with an optimized dictionary.
+
+This type of optimization, combined with the fact that we can use schema awareness at the arrow level to further improve
+the compression ratio, are the main motivations for this design. If we rely solely on the standard collector compression
+configuration, the level of awareness and optimization options will not be the same.
+
+> Note: By storing Arrow buffers in a protobuf field of type 'bytes' we can leverage the zero-copy capability of some
+> Protobuf implementations (e.g. C++, Java, Rust) in order to get the most out of Arrow (relying on zero-copy ser/deser
+> framework).
+
+On the egress stream, a `BatchStatus` message is a collection of `StatusMessage`. A `StatusMessage` is composed of 5
+attributes. The protobuf definition is:
+
+```protobuf
+message BatchStatus {
+  repeated StatusMessage statuses = 1;
+}
+
+message StatusMessage {
+  string batch_id = 1;
+  StatusCode status_code = 2;
+  ErrorCode error_code = 3;
+  string error_message = 4;
+  RetryInfo retry_info = 5;
+}
+
+enum StatusCode {
+  OK = 0;
+  ERROR = 1;
+}
+
+enum ErrorCode {
+  UNAVAILABLE = 0;
+  INVALID_ARGUMENT = 1;
+}
+
+message RetryInfo {
+  int64 retry_delay = 1;
```

Comment:
What happens if different retry_delays are received in one BatchStatus? This appears unnecessary and is a complication for the client to honor. Why not just one RetryInfo at the top level in BatchStatus?

Reply:
I think we could remove the "repeated StatusMessage statuses" field and have a single status message. I don't see any reason to batch status responses, it will only complicate things, and so far I've made a 1:1 correspondence between plog.RequestLogs,ptrace.RequeatSpans,pmetric.RequestMetrics object and BatchStatus object.

Reply:
https://github.com/f5/otel-arrow-adapter/pull/184